gips/0081-indexing-payments.md

```diff
@@ -0,0 +1,165 @@
+
+Putting all of this together the solution consists of the following:
+
+- Indexers disclose publicly their price per unit of work.
+- Gateways send indexing agreements to indexers they want to work with.
+- Indexers accept (on-chain) agreements that meet their criteria and get to work indexing.
+- On a regular basis:
+  - Indexers collect for their work, by posting on-chain the original agreement, a Proof of Indexing (PoI) and the amount of work done for that period.
+  - Gateways monitor the QoS per open agreement and assess if they need to cancel (on-chain) and/or replace any of them.
+
+This solution aims to:
+
+- compensate indexers for the exact amount of work they do.
+- allow indexers to gracefully exit the agreement if it can't be fully fulfilled.
+- cap the amount payers will spend on indexing. Though they might spend but get no service back if the cap is too low.
+- incentivize sufficient QoS by awarding agreements to indexers that provide it.
+- provide sufficient transparency and flexibility for gateways, so that they can abstract the complexity and provide simpler pricing to subgraph developers
+
+### Indexing Indexer Selection Algorithm (IISA)
+
+As mentioned above, gateways get to choose who they work with. In Horizon, gateways can be run be any protocol participant and they get to decide how to implement the indexer selection. Nevertheless we propose that a good IISA needs to considers QoS, available stake (with delegation), and other parameters to find the optimal balance between QoS, load-balancing, economic security and price.
+
+### Arbitration and disputes
+
+Since collections are tied to "amount of work done" and this is reported by the indexers themselves, there needs to be a mechanism to punish malicious reporting. We propose here to use slashing. For this to be fair we need the "amount of work done" to be verifiable by an arbitrator in the case of a dispute.
+
+### Amount of work done
+
+As an initial iteration of the "amount of work done" proxy we propose to use indexed blocks and subgraph entities. In particular the amount to be collected by indexers shall be equal to:
+
+- A price per amount of blocks indexed (with a different price for each chain). These are charged incrementally, i.e. every time a PoI is posted, the indexer collects a payment for the new blocks indexed since the last PoI.
+- A price per amount of entities stored by the subgraph (same price for each chain). These are charged on a recurring basis per unit of time, for the total number of entities on the subgraph at the time the PoI is posted times the time since the last PoI.
+
+We understand that this approach is vulnerable to some forms of abuse. For example, malicious users can deploy subgraphs that are expensive to index but amount to a comparatively small "amount of work done" proxy fee. We suggest that indexers stop serving such subgraphs as soon as possible, to minimize the damage done.
```

Comment:
I would be tempted to add something about us seeking better pricing to address this in the future.

Reply:
The font used by github makes it almost impossible to discern between a I and a l. Does PoI have to be PoI and not POI?

Reply:
> The font used by github makes it almost impossible to discern between a I and a l. Does PoI have to be PoI and not POI?

I actually prefer POI, but picked up PoI from `0009-arbitration-charter.md`, though, to be fair, POI is wide used in various other GIPs, so will switch it up shortly.

Reply:
Not something I impose on you as author, but as an input for you to consider. I prefer capitalisation to match that of what you are abbreviating. As it is written as 'Proof of Indexing', I would do PoI.

My default assumption is alternatives are due to laziness or not having thought about it, rather than design.

Reply:
> I hope developers don't default to putting all their subgraph into a single big entity 😆

This is definitely a risk in the MVP, but worth noting that if they do and it becomes too onerous for Indexers, then they will probably drop the agreements and refuse to serve those subgraphs...
(or if everyone does it, then indexers will raise their prices accordingly).

(We still need to iterate on pricing, but the above is why I think the MVP pricing will work, at least for now)